File: filepicker/src/main/java/droidninja/filepicker/FilePickerBuilder.java
package droidninja.filepicker;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.widget.Toast;
import droidninja.filepicker.models.FileType;
import droidninja.filepicker.models.sort.SortingTypes;
import droidninja.filepicker.utils.Orientation;
import java.util.ArrayList;

/**
 * Created by droidNinja on 29/07/16.
 */
public class FilePickerBuilder {

  private final Bundle mPickerOptionsBundle;

  public FilePickerBuilder() {
    mPickerOptionsBundle = new Bundle();
  }

  public static FilePickerBuilder getInstance() {
    return new FilePickerBuilder();
  }

  public FilePickerBuilder setMaxCount(int maxCount) {
    PickerManager.getInstance().setMaxCount(maxCount);
    return this;
  }

  public FilePickerBuilder setActivityTheme(int theme) {
    PickerManager.getInstance().setTheme(theme);
    return this;
  }

  public FilePickerBuilder setActivityTitle(String title) {
    PickerManager.getInstance().setTitle(title);
    return this;
  }

  public FilePickerBuilder setSelectedFiles(ArrayList<String> selectedPhotos) {
    mPickerOptionsBundle.putStringArrayList(FilePickerConst.KEY_SELECTED_MEDIA, selectedPhotos);
    return this;
  }

  public FilePickerBuilder enableVideoPicker(boolean status) {
    PickerManager.getInstance().setShowVideos(status);
    return this;
  }

  public FilePickerBuilder enableImagePicker(boolean status) {
    PickerManager.getInstance().setShowImages(status);
    return this;
  }

  public FilePickerBuilder enableSelectAll(boolean status) {
    PickerManager.getInstance().enableSelectAll(status);
    return this;
  }

  public FilePickerBuilder setCameraPlaceholder(@DrawableRes int drawable) {
    PickerManager.getInstance().setCameraDrawable(drawable);
    return this;
  }

  public FilePickerBuilder showGifs(boolean status) {
    PickerManager.getInstance().setShowGif(status);
    return this;
  }

  public FilePickerBuilder showFolderView(boolean status) {
    PickerManager.getInstance().setShowFolderView(status);
    return this;
  }

  public FilePickerBuilder enableDocSupport(boolean status) {
    PickerManager.getInstance().setDocSupport(status);
    return this;
  }

  public FilePickerBuilder enableCameraSupport(boolean status) {
    PickerManager.getInstance().setEnableCamera(status);
    return this;
  }

  public FilePickerBuilder withOrientation(Orientation orientation) {
    PickerManager.getInstance().setOrientation(orientation);
    return this;
  }

  public FilePickerBuilder addFileSupport(String title, String[] extensions,
      @DrawableRes int drawable) {
    PickerManager.getInstance().addFileType(new FileType(title, extensions, drawable));
    return this;
  }

  public FilePickerBuilder addFileSupport(String title, String[] extensions) {
    PickerManager.getInstance().addFileType(new FileType(title, extensions, 0));
    return this;
  }

  public FilePickerBuilder sortDocumentsBy(SortingTypes type) {
    PickerManager.getInstance().setSortingType(type);
    return this;
  }

  public void pickPhoto(Activity context) {
    mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.MEDIA_PICKER);
    start(context, FilePickerConst.REQUEST_CODE_PHOTO);
  }

  public void pickPhoto(Fragment context) {
    mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.MEDIA_PICKER);
    start(context, FilePickerConst.REQUEST_CODE_PHOTO);
  }

  public void pickFile(Activity context) {
    mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.DOC_PICKER);
    start(context, FilePickerConst.REQUEST_CODE_DOC);
  }

  public void pickFile(Fragment context) {
    mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.DOC_PICKER);
    start(context, FilePickerConst.REQUEST_CODE_DOC);
  }

  public void pickPhoto(Activity context, int requestCode) {
    mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.MEDIA_PICKER);
    start(context, requestCode);
  }

  public void pickPhoto(Fragment context, int requestCode) {
    mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.MEDIA_PICKER);
    start(context, requestCode);
  }

  public void pickFile(Activity context, int requestCode) {
    mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.DOC_PICKER);
    start(context, requestCode);
  }

  public void pickFile(Fragment context, int requestCode) {
    mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.DOC_PICKER);
    start(context, requestCode);
  }

  private void start(Activity context, int requestCode) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      if (ContextCompat.checkSelfPermission(context, FilePickerConst.PERMISSIONS_FILE_PICKER)
          != PackageManager.PERMISSION_GRANTED) {
        Toast.makeText(context,
            context.getResources().getString(R.string.permission_filepicker_rationale),
            Toast.LENGTH_SHORT).show();
        return;
      }
    }

    PickerManager.getInstance()
        .setProviderAuthorities(
            context.getApplicationContext().getPackageName() + ".droidninja.filepicker.provider");

    Intent intent = new Intent(context, FilePickerActivity.class);
    intent.putExtras(mPickerOptionsBundle);

    context.startActivityForResult(intent, requestCode);
  }

  private void start(Fragment fragment, int requestCode) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      if (ContextCompat.checkSelfPermission(fragment.getContext(),
          FilePickerConst.PERMISSIONS_FILE_PICKER) != PackageManager.PERMISSION_GRANTED) {
        Toast.makeText(fragment.getContext(), fragment.getContext()
            .getResources()
            .getString(R.string.permission_filepicker_rationale), Toast.LENGTH_SHORT).show();
        return;
      }
    }

    PickerManager.getInstance()
        .setProviderAuthorities(fragment.getContext().getApplicationContext().getPackageName()
            + ".droidninja.filepicker.provider");

    Intent intent = new Intent(fragment.getActivity(), FilePickerActivity.class);
    intent.putExtras(mPickerOptionsBundle);

    fragment.startActivityForResult(intent, requestCode);
  }
}

File: filepicker/src/main/java/droidninja/filepicker/FilePickerBuilder.kt
package droidninja.filepicker

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import androidx.annotation.DrawableRes
import androidx.fragment.app.Fragment
import androidx.core.content.ContextCompat
import android.widget.Toast
import androidx.annotation.IntegerRes
import droidninja.filepicker.models.FileType
import droidninja.filepicker.models.sort.SortingTypes
import java.util.ArrayList

/**
 * Created by droidNinja on 29/07/16.
 */
class FilePickerBuilder {

    private val mPickerOptionsBundle: Bundle = Bundle()

    fun setImageSizeLimit(fileSize: Int): FilePickerBuilder {
        PickerManager.imageFileSize = fileSize
        return this
    }

    fun setVideoSizeLimit(fileSize: Int) : FilePickerBuilder{
        PickerManager.videoFileSize = fileSize
        return this
    }

    fun setMaxCount(maxCount: Int): FilePickerBuilder {
        PickerManager.setMaxCount(maxCount)
        return this
    }

    fun setActivityTheme(theme: Int): FilePickerBuilder {
        PickerManager.theme = theme
        return this
    }

    fun setActivityTitle(title: String): FilePickerBuilder {
        PickerManager.title = title
        return this
    }

    /**
     * @param spanType it could be [FilePickerConst.SPAN_TYPE.FOLDER_SPAN] (for folder screen)
     * or [FilePickerConst.SPAN_TYPE.DETAIL_SPAN] (for details screen)
     * @param count span count in integer, defaults for Folder is 2 and Details is 3
     */
    fun setSpan(spanType: FilePickerConst.SPAN_TYPE, count: Int): FilePickerBuilder {
        PickerManager.spanTypes[spanType] = count
        return this
    }

    fun setSelectedFiles(selectedPhotos: ArrayList<Uri>): FilePickerBuilder {
        mPickerOptionsBundle.putParcelableArrayList(FilePickerConst.KEY_SELECTED_MEDIA, selectedPhotos)
        return this
    }

    fun enableVideoPicker(status: Boolean): FilePickerBuilder {
        PickerManager.setShowVideos(status)
        return this
    }

    fun enableImagePicker(status: Boolean): FilePickerBuilder {
        PickerManager.setShowImages(status)
        return this
    }

    fun enableSelectAll(status: Boolean): FilePickerBuilder {
        PickerManager.enableSelectAll(status)
        return this
    }

    fun setCameraPlaceholder(@DrawableRes drawable: Int): FilePickerBuilder {
        PickerManager.cameraDrawable = drawable
        return this
    }

    fun showGifs(status: Boolean): FilePickerBuilder {
        PickerManager.isShowGif = status
        return this
    }

    fun showFolderView(status: Boolean): FilePickerBuilder {
        PickerManager.isShowFolderView = status
        return this
    }

    fun enableDocSupport(status: Boolean): FilePickerBuilder {
        PickerManager.isDocSupport = status
        return this
    }

    fun enableCameraSupport(status: Boolean): FilePickerBuilder {
        PickerManager.isEnableCamera = status
        return this
    }


    fun withOrientation(@IntegerRes orientation: Int): FilePickerBuilder {
        PickerManager.orientation = orientation
        return this
    }

    @JvmOverloads
    fun addFileSupport(title: String, extensions: Array<String>,
                       @DrawableRes drawable: Int = R.drawable.icon_file_unknown): FilePickerBuilder {
        PickerManager.addFileType(FileType(title, extensions, drawable))
        return this
    }

    fun sortDocumentsBy(type: SortingTypes): FilePickerBuilder {
        PickerManager.sortingType = type
        return this
    }

    fun pickPhoto(context: Activity) {
        mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.MEDIA_PICKER)
        start(context, FilePickerConst.REQUEST_CODE_PHOTO)
    }

    fun pickPhoto(context: Fragment) {
        mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.MEDIA_PICKER)
        start(context, FilePickerConst.REQUEST_CODE_PHOTO)
    }

    fun pickFile(context: Activity) {
        mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.DOC_PICKER)
        start(context, FilePickerConst.REQUEST_CODE_DOC)
    }

    fun pickFile(context: Fragment) {
        mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.DOC_PICKER)
        start(context, FilePickerConst.REQUEST_CODE_DOC)
    }

    fun pickPhoto(context: Activity, requestCode: Int) {
        mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.MEDIA_PICKER)
        start(context, requestCode)
    }

    fun pickPhoto(context: Fragment, requestCode: Int) {
        mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.MEDIA_PICKER)
        start(context, requestCode)
    }

    fun pickFile(context: Activity, requestCode: Int) {
        mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.DOC_PICKER)
        start(context, requestCode)
    }

    fun pickFile(context: Fragment, requestCode: Int) {
        mPickerOptionsBundle.putInt(FilePickerConst.EXTRA_PICKER_TYPE, FilePickerConst.DOC_PICKER)
        start(context, requestCode)
    }

    private fun start(context: Activity, requestCode: Int) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (ContextCompat.checkSelfPermission(context, FilePickerConst.PERMISSIONS_FILE_PICKER) != PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(context,
                        context.resources.getString(R.string.permission_filepicker_rationale),
                        Toast.LENGTH_SHORT).show()
                return
            }
        }

        val intent = Intent(context, FilePickerActivity::class.java)
        intent.putExtras(mPickerOptionsBundle)

        context.startActivityForResult(intent, requestCode)
    }

    private fun start(fragment: Fragment, requestCode: Int) {
        fragment.context?.let {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                if (ContextCompat.checkSelfPermission(it,
                                FilePickerConst.PERMISSIONS_FILE_PICKER) != PackageManager.PERMISSION_GRANTED) {
                    Toast.makeText(fragment.context, it
                            .resources
                            .getString(R.string.permission_filepicker_rationale), Toast.LENGTH_SHORT).show()
                    return
                }
            }

            val intent = Intent(fragment.activity, FilePickerActivity::class.java)
            intent.putExtras(mPickerOptionsBundle)

            fragment.startActivityForResult(intent, requestCode)
        }
    }

    companion object {
        @JvmStatic
        val instance: FilePickerBuilder
            get() = FilePickerBuilder()
    }
}
